pkg/flowaggregator/clickhouseclient/clickhouseclient.go

```diff
@@ -0,0 +1,511 @@
+	flowEndSecondsFromSourceNode         time.Time
+	flowEndSecondsFromDestinationNode    time.Time
+	flowEndReason                        uint8
+	sourceIP                             string
+	destinationIP                        string
+	sourceTransportPort                  uint16
+	destinationTransportPort             uint16
+	protocolIdentifier                   uint8
+	packetTotalCount                     uint64
+	octetTotalCount                      uint64
+	packetDeltaCount                     uint64
+	octetDeltaCount                      uint64
+	reversePacketTotalCount              uint64
+	reverseOctetTotalCount               uint64
+	reversePacketDeltaCount              uint64
+	reverseOctetDeltaCount               uint64
+	sourcePodName                        string
+	sourcePodNamespace                   string
+	sourceNodeName                       string
+	destinationPodName                   string
+	destinationPodNamespace              string
+	destinationNodeName                  string
+	destinationClusterIP                 string
+	destinationServicePort               uint16
+	destinationServicePortName           string
+	ingressNetworkPolicyName             string
+	ingressNetworkPolicyNamespace        string
+	ingressNetworkPolicyRuleName         string
+	ingressNetworkPolicyRuleAction       uint8
+	ingressNetworkPolicyType             uint8
+	egressNetworkPolicyName              string
+	egressNetworkPolicyNamespace         string
+	egressNetworkPolicyRuleName          string
+	egressNetworkPolicyRuleAction        uint8
+	egressNetworkPolicyType              uint8
+	tcpState                             string
+	flowType                             uint8
+	sourcePodLabels                      string
+	destinationPodLabels                 string
+	throughput                           uint64
+	reverseThroughput                    uint64
+	throughputFromSourceNode             uint64
+	throughputFromDestinationNode        uint64
+	reverseThroughputFromSourceNode      uint64
+	reverseThroughputFromDestinationNode uint64
+}
+
+func NewClickHouseClient(input ClickHouseInput) (*ClickHouseExportProcess, error) {
+	dsn, err := input.getDataSourceName()
+	if err != nil {
+		klog.Errorf("error parsing ClickHouse DSN: %v", err)
+		return nil, err
+	}
+	connect, err := sql.Open("clickhouse", dsn)
+	if err != nil {
+		klog.Errorf("error when opening DB connection to %s: %v", input.DbURL, err)
+		return nil, err
+	}
+
+	// Test ping DB
+	if err = connect.Ping(); err != nil {
+		if exception, ok := err.(*clickhouse.Exception); ok {
+			klog.Errorf("error pinging DB: [%d] %s", exception.Code, exception.Message)
+		} else {
+			klog.Errorf("error pinging DB: %v", exception)
+		}
+		return nil, err
+	}
+
+	// Test open Transaction
+	tx, err := connect.Begin()
+	if err == nil {
+		_, err = tx.Prepare(insertQuery)
+	}
+	if err != nil {
+		klog.Errorf("error when preparing insert statement: %v", err)
+		return nil, err
+	}
+	_ = tx.Commit()
```

Comment:
Are the prepare or commit operations oing anything against the DB backend, since the TX has not statements? I guess it depends on the backend - not sure how it is with ClickHouse.

Reply:
yes it depends on backend. based on my understanding, the prepare on clickhouse client does actually triggers the insert stmt to be sent and validated to backend db. see: [clickhouse.go](https://github.com/ClickHouse/clickhouse-go/blob/v1/clickhouse.go#L63)